File: cpp/benchmark.cpp
#include <array>
#include <chrono>
#include <cstdint>
#include <iostream>

#if defined(__GNUC__) || defined(__clang__)
#define ALWAYS_INLINE __attribute__((always_inline)) inline
#else
#define ALWAYS_INLINE inline
#endif

static ALWAYS_INLINE uint8_t can_beat_easy_fast(uint8_t a, uint8_t b) {
    uint8_t gt = static_cast<uint8_t>(a > b);
    uint8_t same_type = static_cast<uint8_t>((((a ^ b) & 0xF0u) == 0u));
    uint8_t high_type = static_cast<uint8_t>((a & 0x80u) != 0u);
    return static_cast<uint8_t>(gt & (same_type | high_type));
}

static std::array<uint8_t, 65536> build_table() {
    std::array<uint8_t, 65536> table{};
    for (uint32_t a = 0; a < 256; ++a) {
        uint32_t base = a << 8;
        for (uint32_t b = 0; b < 256; ++b) {
            table[base | b] = can_beat_easy_fast(static_cast<uint8_t>(a), static_cast<uint8_t>(b));
        }
    }
    return table;
}

static uint64_t bench_direct(uint32_t rounds) {
    uint64_t c = 0;
    for (uint32_t r = 0; r < rounds; ++r) {
        for (uint32_t a = 0; a < 256; ++a) {
            for (uint32_t b = 0; b < 256; ++b) {
                c += can_beat_easy_fast(static_cast<uint8_t>(a), static_cast<uint8_t>(b));
            }
        }
    }
    return c;
}

static uint64_t bench_lookup(const std::array<uint8_t, 65536>& table, uint32_t rounds) {
    uint64_t c = 0;
    for (uint32_t r = 0; r < rounds; ++r) {
        for (uint32_t a = 0; a < 256; ++a) {
            uint32_t base = a << 8;
            for (uint32_t b = 0; b < 256; ++b) {
                c += table[base | b];
            }
        }
    }
    return c;
}

int main() {
    constexpr uint32_t rounds = 4000;
    auto table = build_table();

    auto checksum_direct = bench_direct(1);
    auto checksum_lookup = bench_lookup(table, 1);
    std::cout << "correct=" << (checksum_direct == checksum_lookup ? "true" : "false") << "\n";
    std::cout << "true_count=" << checksum_direct << "\n";

    auto s = std::chrono::steady_clock::now();
    auto d = bench_direct(rounds);
    auto e = std::chrono::steady_clock::now();
    double t_direct = std::chrono::duration<double>(e - s).count();

    s = std::chrono::steady_clock::now();
    auto l = bench_lookup(table, rounds);
    e = std::chrono::steady_clock::now();
    double t_lookup = std::chrono::duration<double>(e - s).count();

    std::cout << "direct_checksum=" << d << "\n";
    std::cout << "lookup_checksum=" << l << "\n";
    std::cout << "direct_seconds=" << t_direct << "\n";
    std::cout << "lookup_seconds=" << t_lookup << "\n";
    std::cout << "ratio_direct_over_lookup=" << (t_direct / t_lookup) << "\n";
    return 0;
}

File: java/Benchmark.java
public class Benchmark {
    static int canBeatEasyFast(int a, int b) {
        int gt = (a > b) ? 1 : 0;
        int sameType = ((((a ^ b) & 0xF0) == 0) ? 1 : 0);
        int highType = ((a & 0x80) >>> 7);
        return gt & (sameType | highType);
    }

    static byte[] buildTable() {
        byte[] table = new byte[256 * 256];
        for (int a = 0; a < 256; a++) {
            int base = a << 8;
            for (int b = 0; b < 256; b++) {
                table[base | b] = (byte) canBeatEasyFast(a, b);
            }
        }
        return table;
    }

    static long benchDirect(int rounds) {
        long c = 0L;
        for (int r = 0; r < rounds; r++) {
            for (int a = 0; a < 256; a++) {
                for (int b = 0; b < 256; b++) {
                    c += canBeatEasyFast(a, b);
                }
            }
        }
        return c;
    }

    static long benchLookup(byte[] table, int rounds) {
        long c = 0L;
        for (int r = 0; r < rounds; r++) {
            for (int a = 0; a < 256; a++) {
                int base = a << 8;
                for (int b = 0; b < 256; b++) {
                    c += table[base | b];
                }
            }
        }
        return c;
    }

    public static void main(String[] args) {
        final int warmupRounds = 800;
        final int rounds = 4000;
        byte[] table = buildTable();

        benchDirect(warmupRounds);
        benchLookup(table, warmupRounds);

        long checksumDirect = benchDirect(1);
        long checksumLookup = benchLookup(table, 1);
        System.out.println("correct=" + (checksumDirect == checksumLookup));
        System.out.println("true_count=" + checksumDirect);

        long start = System.nanoTime();
        long d = benchDirect(rounds);
        long end = System.nanoTime();
        double directSeconds = (end - start) / 1_000_000_000.0;

        start = System.nanoTime();
        long l = benchLookup(table, rounds);
        end = System.nanoTime();
        double lookupSeconds = (end - start) / 1_000_000_000.0;

        System.out.println("direct_checksum=" + d);
        System.out.println("lookup_checksum=" + l);
        System.out.printf("direct_seconds=%.6f%n", directSeconds);
        System.out.printf("lookup_seconds=%.6f%n", lookupSeconds);
        System.out.printf("ratio_direct_over_lookup=%.6f%n", directSeconds / lookupSeconds);
    }
}
